src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/CompressedMessageBinaryEncoder.cs

```diff
@@ -0,0 +1,257 @@
+﻿/* Copyright 2013-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using MongoDB.Bson.IO;
+using MongoDB.Driver.Core.Compression;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.WireProtocol.Messages.Encoders.BinaryEncoders
+{
+    /// <summary>
+    /// Represents a binary encoder for a compressed message.
+    /// </summary>
+    public sealed class CompressedMessageBinaryEncoder : MessageBinaryEncoderBase, IMessageEncoder
+    {
+        private readonly ICompressorFactory _compressorFactory;
+        private readonly MessageEncoderSettings _encoderSettings;
+        private const int MessageHeaderLength = 16;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CompressedMessageBinaryEncoder" /> class.
+        /// </summary>
+        /// <param name="stream">The stream.</param>
+        /// <param name="compressorFactory">The compressor factory.</param>
+        /// <param name="encoderSettings">The encoder settings.</param>
+        public CompressedMessageBinaryEncoder(
+            Stream stream,
+            ICompressorFactory compressorFactory,
+            MessageEncoderSettings encoderSettings)
+            : base(stream, encoderSettings)
+        {
+            _compressorFactory = Ensure.IsNotNull(compressorFactory, nameof(compressorFactory));
+            _encoderSettings = encoderSettings;
+        }
+
+        /// <summary>
+        /// Reads the message.
+        /// </summary>
+        /// <returns>A message.</returns>
+        public CompressedMessage ReadMessage()
+        {
+            var stream = CreateBinaryReader().BsonStream;
+            if (stream.Position != 0)
+            {
```

Comment:
maybe we can delegate this responsibility on the user side?

Reply:
I made it to be able to make reading right after the writing without any additional `stream` changes. But I also wasn't confident that it's the right way to do it.
So, I will fix it.